include/boost/geometry/algorithms/detail/envelope/range.hpp

```diff
@@ -35,50 +48,117 @@ namespace detail { namespace envelope
+            >::type iterator_type;
+
+        // initialize MBR
+        initialize<Box, 0, dimension<Box>::value>::apply(mbr);
+
+        // skip through empty geometries
+        iterator_type it = boost::begin(multirange);
+        while (it != boost::end(multirange) && geometry::is_empty(*it))
+        {
+            ++it;
+        }
+
+        // if there are still elements left, apply the envelope policy
+        if (it != boost::end(multirange))
+        {
+            // compute the initial envelope
+            EnvelopePolicy::apply(*it, mbr);
+
+            // for the remaining non-empty geometries just expand the mbr
+            // using the envelope
+            for (++it; it != boost::end(multirange); ++it)
+            {
+                if (! geometry::is_empty(*it))
+                {
+                    Box helper_mbr;
+                    EnvelopePolicy::apply(*it, helper_mbr);
+
+                    dispatch::expand<Box, Box>::apply(mbr, helper_mbr);
+                }
+            }
+        }
     }
-}
+};
 
 
-template <typename ExpandPolicy = default_expand_policy>
-struct envelope_range
+// implementation for multi-range on a spheroid (longitude is periodic)
+template <typename EnvelopePolicy>
+struct envelope_multi_range_on_spheroid
 {
-    /// Calculate envelope of range using a strategy
-    template <typename Range, typename Box>
-    static inline void apply(Range const& range, Box& mbr)
+    template <typename MultiRange, typename Box>
+    static inline void apply(MultiRange const& multirange, Box& mbr)
     {
-        typename boost::range_iterator<Range const>::type it 
-            = boost::begin(range);
-
-        if (it != boost::end(range))
+        typedef typename boost::range_iterator
+            <
+                MultiRange const
+            >::type iterator_type;
+
+        // initialize MBR
+        initialize<Box, 0, dimension<Box>::value>::apply(mbr);
+
+        // compute the boxes of the single geometries
+        std::vector<Box> boxes;
+        for (iterator_type it = boost::begin(multirange);
+             it != boost::end(multirange);
+             ++it)
         {
-            // initialize box with first element in range
-            dispatch::envelope
-                <
-                    typename boost::range_value<Range>::type
-                >::apply(*it, mbr);
+            if (! geometry::is_empty(*it))
+            {
+                Box helper_box;
+                EnvelopePolicy::apply(*it, helper_box);
+                boxes.push_back(helper_box);
+            }
+        }
 
-            // consider now the remaining elements in the range (if any)
-            ++it;
-            envelope_iterators
-                <
-                    ExpandPolicy
-                >(it, boost::end(range), mbr);
+        // compute the envelope of the range of boxes
+        if (! boxes.empty())
+        {
+            envelope_range_of_boxes::apply(boxes, mbr);
```

Comment:
this looks better (single loop).
You could add ```else { initialize<...> }``` instead of the initialization above.